weather_sp/splitter_pipeline/file_splitters.py

```diff
@@ -90,89 +85,132 @@ def should_skip(self):
         with self._copy_to_local_file() as local_file:
             yield pygrib.open(local_file.name)
 
-    def _open_outfile(self, key: SplitKey):
-        return FileSystems.create(self._get_output_file_path(key))
+    def _open_outfile(self, key: str):
+        return FileSystems.create(key)
 
 
 class NetCdfSplitter(FileSplitter):
 
+    def __init__(self, input_path: str, output_info: OutFileInfo,
+                 force_split: bool = False, logging_level: int = logging.INFO):
+        super().__init__(input_path, output_info,
+                         force_split, logging_level)
+        self.output_info.set_formatting_if_needed('_{variable}')
+        self.split_dims = self._get_split_dims()
+
     def split_data(self) -> None:
+        if not self.split_dims:
+            raise ValueError('No splitting specified in template.')
         if self.should_skip():
             metrics.Metrics.counter('file_splitters', 'skipped').inc()
-            self.logger.info('Skipping %s, file already split.', repr(self.input_path))
+            self.logger.info('Skipping %s, file already split.',
+                             repr(self.input_path))
             return
-
-        with self._open_dataset_locally() as nc_data:
-            fields = [var for var in nc_data.variables.keys() if
-                      var not in nc_data.dimensions.keys()]
-            for field in fields:
-                self._create_netcdf_dataset_for_variable(nc_data, field)
-            self.logger.info('split %s into %d files', self.input_path, len(fields))
+        if any(split not in ('time', 'level', 'variable') for split in self.split_dims):
+            raise ValueError(
+                'netcdf split: unknown split dimension, supported are time, level, variable')
+
+        with self._open_dataset_locally() as dataset:
+            if any(split not in dataset.dims and split not in ('variable') for split in self.split_dims):
+                raise ValueError(
+                    'netcdf split: requested dimension not in dataset')
+            iterlists = []
+            iterlists.append([dataset[var].to_dataset(
+            ) for var in dataset.data_vars] if 'variable' in self.split_dims else [dataset])
+            for dim in ('time', 'level'):
+                if dim in self.split_dims:
+                    iterlists.append(dataset[dim])
+            combinations = list(itertools.product(*iterlists))
+            for comb in combinations:
+                selected = comb[0]
+                for da in comb[1:]:
+                    if 'time' in da.coords:
+                        selected = selected.sel(time=da.time)
+                    if 'level' in da.coords:
+                        selected = selected.sel(level=da.level)
+                self._write_dataset(selected)
+            self.logger.info('split %s into %d files',
+                             self.input_path, len(combinations))
```

Comment:
It's a bit late in the day, I'll take a closer look at this block tomorrow. I do think that XArray's indexing and selection capabilities should provide enough affordances for arbitrary selection of chunks of datasets.

Reply:
I was planning to look more into this as well to get this more flexible. If you have some example NetCDF files with more dimensions / coordinates that would be helpful for testing. However, for this first version I would like to leave this and tackle NetCDF improvements in a separate PR.